### ISSUE TITLE
Default color cycle is too short

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#1591]
As part of #1518 I noted that our current color cycle is very short:

![image](https://user-images.githubusercontent.com/890576/27388526-a916c34a-5693-11e7-8c62-3dff2f9da940.png)

For comparison, the matplotlib 2.0 color cycle has 10 colors:

![image](https://user-images.githubusercontent.com/890576/27388550-bbff586e-5693-11e7-84c7-71853fa0bc91.png)

One suggestion is to extend our color cycle by appending the matplotlib purple, brown and any other distinguishable colors.

Another suggestion is to use the large color cycle in datashader that is the union of Set1,Set2 and Set3 (minus colors that are too similar).

![image](https://user-images.githubusercontent.com/890576/27388508-9b98c498-5693-11e7-9988-cb415920e7ca.png)

PersonallyI don't find all these colors quite distinguishable enough so I prefer the first option (extending our current cycle, especially as it is likely to cause less disruption).

**Comment by ea42gh** [comment]
http://colorbrewer2.org/#type=qualitative&scheme=Paired&n=12

**Comment by jlstevens** [comment]
@ea42gh We are considering just extending the existing color cycle. Do you have any particular issues with it (other than length) that would lead you to recommend something else entirely different?

**Comment by jlstevens** [comment]
@jbednar @philippjfr I don't think we ended up doing this for 1.8. Should we try to do it for 1.8.1? And if so, should we pretend we did do it so it can fall under the style_17 compatibility switch?

**Comment by jbednar** [comment]
I think it would fall under style_17 in any case, so it's not pretendig, because that switch needs to restore whatever 1.7 did.  Do you mean, so that we don't also have to make a style_18 switch? Sure.

**Comment by jlstevens** [comment]
>  Do you mean, so that we don't also have to make a style_18 switch? Sure.

Right.

**Comment by philippjfr** [comment]
It also seems like 7 and 12 are fairly close. Here's my suggestion with colors taken from Category10/tab10:

```
['#30a2da', '#fc4f30', '#e5ae38', '#6d904f', '#8b8b8b', '#9467bd',  '#8c564b', '#e377c2', '#1f77b4', '#d62728', '#17becf', '#bcbd22']
```

<img width="312" alt="screen shot 2018-03-22 at 12 47 23 pm" src="https://user-images.githubusercontent.com/1550771/37771337-34fef7a2-2dcf-11e8-973d-e04bb88b7e60.png">


Adag

**Comment by jlstevens** [comment]
Looks good! I think I would swap the last one (i.e 12) with the brown (7) to try and balance light and dark colors out a bit.

Otherwise, I would be happy to merge such a change. Certainly a *lot* better than what we had before!

**Comment by jlstevens** [comment]
Do we still want to control this change behind a config switch so people can go back to the short color cycle to make sure their plots don't suddenly look different?

**Comment by jbednar** [comment]
I have no objection, apart from wanting the cycle never to end, and thus to have as many colors *after* the ones here as are even remotely distinguishable, but I know you two don't agree...

**Comment by philippjfr** [comment]
Trying to find a principled way of ordering, e.g. by maximizing distance in lab color space. Think that's going to be worth it?

**Comment by jlstevens** [comment]
> I have no objection, apart from wanting the cycle never to end...

Even 16 million colors isn't *quite* infinite. ;-p

**Comment by jlstevens** [comment]
> Trying to find a principled way of ordering, e.g. by maximizing distance in lab color space. Think that's going to be worth it?

I wouldn't spend too much time on it but it might be worth having a go in case it results in a clearly superior looking cycle. That said the first 5 colors are still fixed for backwards compatibility...

**Comment by philippjfr** [comment]
Wrote a little visualization that might help:

<img width="1013" alt="screen shot 2018-03-22 at 1 12 40 pm" src="https://user-images.githubusercontent.com/1550771/37772502-bc8de95a-2dd2-11e8-8f5e-46d7101ac478.png">

**Comment by jlstevens** [comment]
Great! The problem is the first 5 entries are going down on ``y`` and those are ones we don't want to change.

**Comment by philippjfr** [comment]
Right the smallest transition is from green to grey, which we are forced to keep for now.

**Comment by jbednar** [comment]
Be careful to use "fire" rather than "hot" in the one on the left if you want to draw valid conclusions. :-)

It's certainly possible to use a principled approach, e.g. taking a perceptually uniform color space and sampling it in a fractal-like way that chooses the first five most distinguishable colors, then fills in between them, then fills in between those, etc., gradually adding less discriminable colors.  But even that isn't likely to be valid, because the perceptually uniform color spaces I know of are defined using large color patches that don't really apply to plotting lines or points. Our ability to detect the color of isolated thin lines or points varies quite dramatically depending on size. So really the only principled way to do it would be to run some psychophysical experiments testing discriminability for small dot sizes and thin lines, which there may be some research on but isn't accounted for in the major uniform color spaces (Lab, Luv, etc.).  And it would be nice to run such testing on color blind people as well, e.g. so that the first five colors are color-blind safe, and then the rest relax that criterion (as making the entire cycle colorblind safe would drastically limit the number of colors available).  That all adds up to a major research project, which I think someone should do, but which I don't personally have funding for...

**Comment by philippjfr** [comment]
Haha, the distance matrix is impossible to read anyway, so I'm mostly judging on the difference curve. Switching 7 and 12 as Jean-Luc suggested seems reasonable:

<img width="1052" alt="screen shot 2018-03-22 at 1 26 43 pm" src="https://user-images.githubusercontent.com/1550771/37773265-0fd6cb34-2dd5-11e8-901d-9e35dfdea2fa.png">

**Comment by jbednar** [comment]
The one thing you can read clearly from the matrix is that color 5 (gray) is hard to distinguish from *everything*.  It would be a breaking change to get rid of it, but I do think that particular color is a liability.

**Comment by philippjfr** [comment]
Last variant I'll post:

<img width="1024" alt="screen shot 2018-03-22 at 1 40 00 pm" src="https://user-images.githubusercontent.com/1550771/37773864-a2030652-2dd6-11e8-957d-196f4d173e3f.png">

**Comment by jlstevens** [comment]
The problem with the last one is that when it cycles, you'll have two blues next to each other.

**Comment by philippjfr** [comment]
>The problem with the last one is that when it cycles, you'll have two blues next to each other.

Good point, should have included the cycle.

**Comment by philippjfr** [comment]
Okay last one:

```
colors = ['#30a2da', '#fc4f30', '#e5ae38', '#6d904f', '#8b8b8b', '#17becf',
          '#9467bd', '#d62728', '#1f77b4', '#e377c2', '#8c564b', '#bcbd22']
```

<img width="1031" alt="screen shot 2018-03-22 at 1 49 35 pm" src="https://user-images.githubusercontent.com/1550771/37774368-e67bf8e2-2dd7-11e8-9a85-d71ca7e850d0.png">

Guess we should vote or something?

**Comment by jbednar** [comment]
The last one there seems fine to me, except that I vote to remove gray entirely as not being a color and being hard to distinguish from all colors.

**Comment by jlstevens** [comment]
I think removing the gray should be an issue for 2.0 as removing it breaks backwards compatibility. Otherwise, I agree that it should be removed.

**Comment by philippjfr** [comment]
The extended cycle is now, merged, moving to the 2.0 milestone to eventually replace the color cycle entirely.
